src/main/kotlin/solve/scene/controller/SceneController.kt

```diff
@@ -49,42 +90,98 @@ class SceneController : Controller() {
+            reinitializeSettings(newScene)
+        } else {
+            recalculateScale()
         }
     }
 
-    fun zoomIn(mousePosition: DoublePoint) = zoom(min(scaleProperty.value * scaleFactor, MaxScale), mousePosition)
+    fun zoomIn(mousePosition: DoublePoint) = zoom(min(scale * ScaleFactor, maxScale), mousePosition)
+
+    fun zoomOut(mousePosition: DoublePoint) = zoom(max(scale / ScaleFactor, min(minScale, scale)), mousePosition)
 
-    fun zoomOut(mousePosition: DoublePoint) =
-        zoom(max(scaleProperty.value / scaleFactor, min(minScale, scaleProperty.value)), mousePosition)
+    fun recalculateScale() {
+        scale = max(minScale, calculateMinScaleDependingOnColumns())
+    }
 
     private fun zoom(newScale: Double, mousePosition: DoublePoint) {
-        val initialMouseX = (xProperty.value + mousePosition.x) / scaleProperty.value
-        val initialMouseY = (yProperty.value + mousePosition.y) / scaleProperty.value
+        val initialMouseX = (xProperty.value + mousePosition.x) / scale
+        val initialMouseY = (yProperty.value + mousePosition.y) / scale
 
         scaleProperty.value = newScale
 
-        x = initialMouseX * scaleProperty.value - mousePosition.x
-        y = initialMouseY * scaleProperty.value - mousePosition.y
+        x = initialMouseX * scale - mousePosition.x
+        y = initialMouseY * scale - mousePosition.y
+    }
+
+    private fun reinitializeSettings(newScene: Scene) {
+        columnsNumber = calculateColumnsCount(newScene)
+        setDefaultScaleRange()
+        scale = calculateMinScaleDependingOnColumns()
+    }
+
+    private fun addGridSettingsBindings() {
+        columnsNumberProperty.onChange { newColumnsNumber ->
+            newColumnsNumber ?: return@onChange
+
+            scale = calculateMinScaleDependingOnColumns()
+        }
+
+        installedMinScaleProperty.onChange { installedMinScale ->
+            if (installedMinScale > scale) {
+                scale = installedMinScale
+            }
+        }
+
+        installedMaxScaleProperty.onChange { installedMaxScale ->
+            if (installedMaxScale < scale) {
+                scale = installedMaxScale
+            }
+        }
+    }
+
+    private fun calculateMinScaleDependingOnColumns(): Double {
+        val displayingColumnsNumber = min(scene.frames.count(), columnsNumber)
+
+        return min(
+            max(
+                installedMinScale,
+                sceneWidthProperty.value /
+                        ((scene.frameSize.width + SceneView.framesMargin) * displayingColumnsNumber)
+            ),
+            DefaultMaxScale
+        )
+    }
+
+    private fun setDefaultScaleRange() {
+        installedMaxScale = DefaultMaxScale
+        installedMinScale = DefaultMinScale
     }
 
     companion object {
-        private const val defaultX = 0.0
-        private const val defaultY = 0.0
+        const val DefaultMinScale = 0.2
+        const val DefaultMaxScale = 20.0
```

Comment:
Not synchronized with settings
![image](https://user-images.githubusercontent.com/55746901/232112097-ef76a655-ac99-4436-a4bb-ff3ee0e90e40.png)

Reply:
It's strange, at my machine it always show the correct value (20.0), as in the gif